### PR TITLE
Improve loclass logic for readers doing keyrolling.

### PR DESCRIPTION
**Pull request by nvx**
Improve loclass logic for readers doing keyrolling. Writing to the loclass log is also now deferred until the CSN changes for speed.

# What's new

Removes the forced anticoll between MAC collections except for when the CSN changes. This speeds up the collection process and should be more reliable with readers doing keyrolling.

# Verification 

- Perform loclass and ensure it is still able to recover keys from a legacy elite keyed reader.

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: picopass/picopass_worker.c [hunk 1]
```diff
@@ -1035,25 +1052,40 @@ static void picopass_emu_handle_packet(
                 FURI_LOG_W(TAG, "loclass: standard key detected during collection");
                 ctx->loclass_got_std_key = true;
 
-                ctx->state = PicopassEmulatorStateIdle;
+                // Don't reset the state in case the reader tries a non-standard key next without a new
```
Comment (bettse) [inline picopass/picopass_worker.c]: This sentence might be incomplete

File: picopass/picopass_device.h [hunk 2]
```diff
@@ -110,6 +117,7 @@ typedef struct {
     uint8_t key_block_num; // in loclass mode used to store csn#
     bool loclass_mode;
     bool loclass_got_std_key;
+    uint8_t loclass_mac_buffer[8 * LOCLASS_NUM_PER_CSN];
```
Comment (bettse) [inline picopass/picopass_device.h:121]: is this 8 because that is the mac length?

Comment (nvx) [reply picopass/picopass_device.h:121]: 4 bytes for NR plus 4 bytes for MAC, yeah

File: picopass/picopass_worker.c (hunk 1 above)
Comment (nvx) [reply picopass/picopass_worker.c]: Yup, fixed.

**Review by bettse — DISMISSED** [review]
I don't claim to know all the details of loclass, but overall the codes makes sense.

**Comment by bettse** [comment]
I used the nfc-iclass project to make a "KDE" keyroll config card, which I used on a Rev A reader.  I ran the new code against it, downloaded the file, and was able to use https://loclass.ericbetts.dev/ to calculate the key I'd configured :grin:.

**Comment by nvx** [comment]
Fixed that bug with the numbering in the loclass log being wrong and also made the code a little more defensive in the emulation loop to stop when finished rather than relying on the UI to stop the emulation.

**Comment by bettse** [comment]
Pulled down the new code and used it again with my keyrolled reader.  So fast!  Log looks good and works with loclass.ericbetts.dev.  I'm going to use my new found poweres to approve try the "Approve and run" in the PR.

**Review by bettse — APPROVED** [review]

**Comment by bettse** [comment]
Looks like I can approve, but merging is blocked 👍

**Comment by skotopes** [comment]
@bettse looks like you are missing in codeowners. Can you make PR with it?

**Review by skotopes — APPROVED** [review]

**Comment by skotopes** [comment]
Also don't forget changelog + catalog update

**Comment by bettse** [comment]
We've also got https://github.com/flipperdevices/flipperzero-good-faps/pull/47 open, so we could do the bump after that

**Comment by skotopes** [comment]
okok

**Pull request merged**